### PR TITLE
Add support for ConnectTimeout and RecvTimeout

### PR DESCRIPTION
**Pull request by wchen-r7**
This only modifies msf and rex, no modules are modified. If these changes are okay, then we can move on to modules.
# Verification Steps
- [ ] Save this file as auxiliary/gather/test.rb: https://gist.github.com/wchen-r7/164c40de6cc8a01e5677
## Test connection timeout
- [ ] Set up an HTTP server (I did it with IIS)
- [ ] Block that HTTP port (firewall)
- [ ] In test.rb, make sure the `try_user_defined_timeouts` call is uncommented.
- [ ] In test.rb, make sure the `try_mod_defined_timeouts` cals is commented out.
- [ ] Start msfconsole
- [ ] `use auxiliary/gather/test`
- [ ] `set rhost [IP]`
- [ ] `set ConnectTimeout 3`
- [ ] `run`
- [ ] You should see that the Time spent message is about 3 seconds. For example:

```
msf auxiliary(http_test) > run

[*] Your connection timeout is: 3
[*] Your response timeout is  : 5
[*] Sending request
[-] Rex::ConnectionTimeout The connection timed out (192.168.1.169:8181).
[*] Time spent: 3.062065
[*] Your response object is: NilClass
[*] Auxiliary module execution completed
```
## Test response timeout
- [ ] Set up a netcat server as a fake HTTP server: `nc -l 8181`
- [ ] Start msfconsole
- [ ] `use auxiliary/gather/test`
- [ ] `set rhost [IP]`
- [ ] `set RecvTimeout 1`
- [ ] `run`
- [ ] You should see that the Time spent message is about 1 second. For example:

```
msf auxiliary(http_test) > run

[*] Your connection timeout is: 3
[*] Your response timeout is  : 1
[*] Sending request
[*] Time spent: 1.227949
[*] Your response object is: NilClass
[*] Auxiliary module execution completed
```
## Test HTTP authentication
- [ ] Preferably IIS (I did it on a Windows 8.1 + IIS box)
- [ ] Make sure your IIS supports: Basic auth and Windows auth.
- [ ] Create a folder called "auth" in C:\Inetpub\wwwroot\
- [ ] Open IIS Manager
- [ ] Find the "auth" folder under Default Web Site, click on that.
- [ ] On the right, click on Authentication
- [ ] Disable Anonymous Authentication
- [ ] Enable Windows Authentication
- [ ] Open a browser, and make sure http://[server ip]/auth/ is asking you to auth
- [ ] Start msfconsole
- [ ] `use auxiliary/gather/test`
- [ ] `use rhost [ip]`
- [ ] `set targeturi /auth/`
- [ ] `set username [valid username]`
- [ ] `set password [bad password]`
- [ ] `run`
- [ ] You should get a 401
- [ ] Try again with a valid password
- [ ] You should NOT get a 401.
- [ ] Go back to IIS Manager: disable Windows Auth, and enable Basic Auth.
- [ ] Try the test module again against Basic Auth.

### CODE REVIEW

File: lib/msf/core/exploit/http/client.rb [hunk 1]
```diff
@@ -194,7 +196,9 @@ def connect(opts={})
       'SendSPN'                => datastore['NTLM::SendSPN'],
       'UseLMKey'               => datastore['NTLM::UseLMKey'],
       'domain'                 => datastore['DOMAIN'],
-      'DigestAuthIIS'          => datastore['DigestAuthIIS']
+      'DigestAuthIIS'          => datastore['DigestAuthIIS'],
+      'connect_timeout'        => datastore['ConnectTimeout'],
+      'receive_timeout'        => datastore['RecvTimeout']
```
Comment (jvazquez-r7) [inline lib/msf/core/exploit/http/client.rb]: It's called ```receive_timeout``` here but on ```Rex::Proto:Http::Client#initialize``` looks like it has been defined as ```recv_timeout```:

```ruby
    self.config_types = {
      # ... blah blah
      'recv_timeout'           => 'integer'
    }

```

Maybe I'm missing something but I'd say it should be ```recv_timeout``` also here?

File: lib/msf/core/exploit/http/client.rb [hunk 2]
```diff
@@ -254,7 +258,9 @@ def configure_http_login_scanner(conf)
       ntlm_send_spn:                 datastore['NTLM::SendSPN'],
       ntlm_use_lm_key:               datastore['NTLM::UseLMKey'],
       ntlm_domain:                   datastore['DOMAIN'],
-      digest_auth_iis:               datastore['DigestAuthIIS']
+      digest_auth_iis:               datastore['DigestAuthIIS'],
+      connect_timeout:               datastore['ConnectTimeout'],
+      receive_timeout:               datastore['receive_timeout']
```
Comment (jvazquez-r7) [inline lib/msf/core/exploit/http/client.rb]: I'd say the datastore option has been defined as ```RecvTimeout``` not ```receive_timeout```.

Comment (jvazquez-r7) [inline lib/msf/core/exploit/http/client.rb]: Also, maybe has sense to use ```recv_tiemout``` as key for the login_scanner conf, so we use the same key on every place.

File: lib/msf/core/exploit/http/client.rb (hunk 1 above)
Comment (wchen-r7) [inline lib/msf/core/exploit/http/client.rb]: Ah yes. Was changing this name too many times.

BTW what do you like:

1. recv_timeout, or
2. res_timeout

?

Comment (jvazquez-r7) [inline lib/msf/core/exploit/http/client.rb]: recv_timeout looks good to me!

File: lib/rex/proto/http/client.rb [hunk 3]
```diff
@@ -211,10 +214,15 @@ def close
   # authentication and return the final response
   #
   # @return (see #_send_recv)
-  def send_recv(req, t = -1, persist=false)
-    res = _send_recv(req,t,persist)
+  def send_recv(opts={})
+    req             = opts['request']
```
Comment (jvazquez-r7) [inline lib/rex/proto/http/client.rb]: You're initializing ```req```, ```connect_timeout```, ```recv_timeout``` and ```persist``` local variables, but I don't think they are used at all. Do you need them for anything in the future?

File: lib/rex/proto/http/client.rb [hunk 4]
```diff
@@ -255,11 +267,17 @@ def send_request(req, t = -1)
...
-  # @param t [Fixnum] the timeout for the request in seconds
+  # @param connect_timeout [Fixnum] Default: 5
+  # @param recv_timeout [Fixnum] Default: 5
   # @param persist [Boolean] whether or not to persist the TCP connection (pipelining)
   #
   # @return [Response] the last valid HTTP response object we received
-  def send_auth(res, opts, t, persist)
+  def send_auth(opts={})
+    res = opts['response']
+    connect_timeout = opts['connect_timeout'] || 5
```
Comment (jvazquez-r7) [inline lib/rex/proto/http/client.rb]: Your are initializing connect_timeout, recv_timeout and persist local variables, but I don't think they're being used in the method.

File: lib/msf/core/exploit/http/client.rb (hunk 1 above)
Comment (wchen-r7) [inline lib/msf/core/exploit/http/client.rb]: OK

File: lib/rex/proto/http/client.rb (hunk 4 above)
Comment (jvazquez-r7) [inline lib/rex/proto/http/client.rb]: Also have into account the initial values are not used at all. So even when you are initializing connect_timeout, you're not forwarding it to ```_send_recv``` unless it was already in the ```opts``` hash

File: lib/rex/proto/http/client.rb [hunk 5]
```diff
@@ -319,7 +337,8 @@ def basic_auth_header(username,password)
   def digest_auth(opts={})
     @nonce_count = 0
 
-    to = opts['timeout'] || 20
+    connect_timeout = opts['connect_timeout'] || 5
```
Comment (jvazquez-r7) [inline lib/rex/proto/http/client.rb]: Even when you're initializing connect_timeout and recv_timeout you're not using them.

File: lib/rex/proto/http/client.rb [hunk 6]
```diff
@@ -466,7 +485,8 @@ def negotiate_auth(opts={})
       :send_ntlm        => self.config['send_ntlm']
     }
 
-    to = opts['timeout'] || 20
+    connect_timeout = opts['connect_timeout'] || 5
```
Comment (jvazquez-r7) [inline lib/rex/proto/http/client.rb]: Even when you're initializing connect_timeout and recv_timeout you're not using them.

File: lib/rex/proto/http/client.rb [hunk 7]
```diff
@@ -229,10 +237,14 @@ def send_recv(req, t = -1, persist=false)
   # authentication handling.
   #
   # @return (see #read_response)
-  def _send_recv(req, t = -1, persist=false)
-    @pipeline = persist
-    send_request(req, t)
-    res = read_response(t)
+  def _send_recv(opts={})
```
Comment (jvazquez-r7) [inline lib/rex/proto/http/client.rb]: I miss documentation for the ```opts``` Hash here.

File: lib/msf/core/exploit/http/client.rb [hunk 8]
```diff
@@ -52,7 +52,9 @@ def initialize(info = {})
         OptBool.new('SSL', [ false, 'Negotiate SSL for outgoing connections', false]),
         OptEnum.new('SSLVersion', [ false, 'Specify the version of SSL that should be used', 'Auto', ['Auto', 'SSL2', 'SSL3', 'TLS1']]),
         OptBool.new('FingerprintCheck', [ false, 'Conduct a pre-exploit fingerprint verification', true]),
-        OptString.new('DOMAIN', [ true, 'The domain to use for windows authentification', 'WORKSTATION'])
+        OptString.new('DOMAIN', [ true, 'The domain to use for windows authentification', 'WORKSTATION']),
+        OptInt.new('ConnectTimeout', [ true, 'The amount of time in seconds to establish the connection', 5 ]),
+        OptInt.new('RecvTimeout', [ true, 'The amount of time to wait for server response before closing the connection', 5 ])
```
Comment (jvazquez-r7) [inline lib/msf/core/exploit/http/client.rb]: Finally using a default Timeout of 5 for Recv? Maybe we already talked about it last week and I don't remember. I thought the last conclusion was default 5 for connect and default 20 for Recv. Makes more sense to have 5 and 5 as default?

**Pull request closed without merging**